lib/cocoapods/resolver.rb

```diff
@@ -395,6 +400,27 @@ def handle_resolver_error(error)
       raise Informative, error.message
     end
 
+    # Returns whether the given spec is platform-compatible with the dependency
+    # graph, taking into account the dependency that has required the spec.
+    #
+    # @param  [Molinillo::DependencyGraph] dependency_graph
+    #
+    # @param  [Dependency] dependency
+    #
+    # @param  [Specification] specification
+    #
+    # @return [Bool]
+    def spec_is_platform_compatible?(dependency_graph, dependency, spec)
+      all_predecessors = ->(vertex) do
+        pred = vertex.predecessors
+        pred + pred.map(&all_predecessors).reduce(Set.new, &:&) << vertex
+      end
+      vertex = dependency_graph.vertex_named(dependency.name)
+      predecessors = all_predecessors[vertex].reject { |v| v.explicit_requirements.empty? }
+      platforms_to_satisfy = predecessors.flat_map(&:explicit_requirements).flat_map { |r| @platforms_by_dependency[r] }
+      platforms_to_satisfy.all? { |pts| spec.available_platforms.select { |p| p.name == pts.name }.all? { |p| pts.supports?(p) } }
```

Comment:
I’d personally like to see this spread over a few more lines and I’m not a fan of too many short variable names.